Add support for Dyson HP04 Model

Parlane opened the pull request:
HP04 is an air purifier and and heater/fan. HP04 uses v2 of the api.

Parlane commented:
This fixes issue #4

Parlane commented:
Sorry, this needs more work to make it compatible with homeassistant.

Parlane commented:
`
AttributeError: 'DysonEnvironmentalSensorV2State' object has no attribute 'dust'

AttributeError: 'DysonPureHotCoolV2State' object has no attribute 'filter_life'

AttributeError: 'DysonPureHotCoolV2State' object has no attribute 'focus_mode'

AttributeError: 'DysonPureHotCoolV2State' object has no attribute 'fan_mode'
`

etheralm commented:
Hi, thanks a  lot for the pr, looks good on a first glance. Sorry about the late response I just came back from vacation. I'll have deeper look when I get the time and I'll get back to you.

itssimon commented:
I have two HP04 devices at home in case you need any help with testing!

ryanrhughes commented:
I'm not sure if it's helpful for review or not but I managed to get @Parlane's update working with the Home Assistant component. https://github.com/ryanrhughes/home-assistant/commit/5533ade0e60f41ad2f091af76c0ba4788c468394

etheralm commented on libpurecool/dyson_pure_state_v2.py:
```diff
@@ -241,3 +296,71 @@ def __repr__(self):
+        self._tilt = DysonPureCoolV2State._get_field_value(self._state, 'tilt')
+
+        self._heat_target = DysonPureCoolV2State._get_field_value(self._state,
+                                                                'hmax')
```
Ident is wrong, please align it with 'self' on the line above

etheralm commented on libpurecool/dyson_pure_state_v2.py:
```diff
@@ -241,3 +296,71 @@ def __repr__(self):
+                                                                'hmax')
+
+        self._heat_mode = DysonPureCoolV2State._get_field_value(self._state,
+                                                              'hmod')
```
Same as above

etheralm commented on libpurecool/dyson_pure_state_v2.py:
```diff
@@ -241,3 +296,71 @@ def __repr__(self):
+                                                              'hmod')
+
+        self._heat_state = DysonPureCoolV2State._get_field_value(self._state,
+                                                               'hsta')
```
Same as above

etheralm commented on libpurecool/dyson_pure_state_v2.py:
```diff
@@ -47,6 +54,14 @@ def front_direction(self):
+        """Focus the fan on one stream or spread. - Backwards compat with non-v2"""
+        if self._front_direction == FrontalDirection.FRONTAL_ON.value:
+            return FocusMode.FOCUS_ON.value
+        else:
```
Please remove the else.

etheralm commented on libpurecool/dyson_pure_state_v2.py:
```diff
@@ -72,6 +87,18 @@ def continuous_monitoring(self):
+            return FanMode.FAN.value
+        elif self._speed == FanSpeed.FAN_SPEED_AUTO.value:
+            return FanMode.AUTO.value
+        else:
```
Please remove the last else here

etheralm commented:
> 
> 
> I'm not sure if it's helpful for review or not but I managed to get @Parlane's update working with the Home Assistant component. [ryanrhughes/home-assistant@5533ade](https://github.com/ryanrhughes/home-assistant/commit/5533ade0e60f41ad2f091af76c0ba4788c468394)

@ryanrhughes thx, I just tested his changes with HA and the fan model that I have (DP04), just to be sure and it seems to work fine. Glad to hear that HA is working also with HP04. 

I'll try to push the new version to pypi until the end of the week and then to get it working officially in ha afterwards.

Parlane commented:
Give me until the end of the week to fix this up as there were some issues still integrating with home assistant. Mostly that the climate control doesn't apply setting changes. And I possibly broke backwards compatibility with other v2 devices.

etheralm commented:
np take your time, let me know if I can help with anything.

lwis commented:
@Parlane anything I can do to help you get this merged? I know a few people are keen to get HP04 support in HA.

Parlane commented:
Looking at it now.

Parlane commented:
Fixed the code review notes, but working on the HA integration issues now.

Parlane commented:
With the current changes it does seem the climate device "works, but if you compare it to my heat pump you can see it doesn't support the OFF mode. In the image below it says it is cooling but it is actually turned off currently.

![image](https://user-images.githubusercontent.com/86614/63206922-dbef2b00-c111-11e9-8dbd-a1a4cd83f739.png)

Parlane commented:
The way I am testing this is by going in to the custom_components directory and doing:

git clone https://github.com/Parlane/ha_dyson.git dyson

This overrides the built in dyson component.

Parlane commented:
OK under https://github.com/Parlane/ha_dyson/commit/66cdab975fba0d854ab2956d7cfdc6ee5acc2d73 I have made the changes required in home assistant to support this work.

The only worry I have is that I needed to add fpwr to DysonPureHotCoolLink, which I worry will break pre-hp04 dyson devices. Maybe it might need another subclass just for hp04 ?

I only have an hp04 to test with.

lwis commented:
@Parlane Just tested using your custom component on my HP04;

- Fan isn't working, I can set the fan speed, but I'm unable to turn it on/change oscillation/change to auto.
- Climate looks good now, can switch hot/cool/off perfectly.
- Are you seeing VoC/p10/p25/NO2 in HA?

Parlane commented:
I will have a look at the fan module tonight, I've never tried it to be honest...

The NO2 etc are not added to sensor.py: https://github.com/Parlane/ha_dyson/blob/master/sensor.py

These could be added as additional sensors.

There is dust support existing which I added to EnvironmentV2, it uses max(self._particulate_matter_25, self._particulate_matter_10) provided by DysonDustSensor

VoC is provided by DysonAirQualitySensor

I tried to make the least changes to HA as possible just to make it easier to test, thats why I added a lot of backwards compat functions such as dust :)

Parlane commented:
![image](https://user-images.githubusercontent.com/86614/63229755-fd5a2f00-c257-11e9-8ee3-2a63e5192cbf.png)

lwis commented:
@Parlane gotcha, I see that the data is exposed from the lib - we can enhance HA in the future.

honeyworld2 commented:
> @Parlane Just tested using your custom component on my HP04;
> 
> * Fan isn't working, I can set the fan speed, but I'm unable to turn it on/change oscillation/change to auto.
> * Climate looks good now, can switch hot/cool/off perfectly.
> * Are you seeing VoC/p10/p25/NO2 in HA?

Hi all, thanks for your great work! Any progress with turning on/off the Fan / Auto Mode... ?

canthefason commented:
These two PRs will fix the fan auto mode issue. PTAL.

https://github.com/Parlane/libpurecool/pull/1
https://github.com/Parlane/ha_dyson/pull/1

brott8 commented:
@Parlane , looks great, thanks for the work you put into it.
Just a quick one, I think your Air Quality sensor is referencing the incorrect environmental_state.  "volatil_organic_compounds" should be the VOC sensor.

I guess the AQI would be the max of the other sensors as defined in air_quality.py.
I'll give it a try later.

edit: I realised that is in the base code. Not sure if that was intended. Anyway, that would be a reason the AQI doesn't agree with the value on the device (at least for the HP04)

FutureGUIs commented:
Any extra help needed here?  I just got a hp04 and really would like this PR pushed through.  I've got 2 hp02's as well if we just need testing and/or coding help too.

etheralm commented:
Hey everyone, sorry that this is taking so long. I have a bit of a free time right now and I intend to get this finally done. 

I've created a new branch [here](https://github.com/etheralm/libpurecool/tree/new_hot_cool).  You can track the progress there. So far I haven't made any changes but I plan to cobble up an initial version tomorrow or on Saturday. I'll also try to create a short script to test if everything works properly which I would ask you to run since I don't have that model and can't test it myself. Stay tuned.

etheralm commented:
OK I think I'm ready with the needed changes.  I've posted instructions how to test them in the issue ticket https://github.com/etheralm/libpurecool/issues/4

I'll close this PR for now, lets continue the discussion there as to not have it in multiple places at the same time.

FutureGUIs commented:
Well, I'm not all up on python that much yet, but my regular hp02 devices seem to connect and send/get data just fine from the logging I see, but my hp04 doesn't even connect, even trying to manually connect using its IP address...  I get no logging either with the hp04
